### PR TITLE
add dep-check task

### PR DESCRIPTION
**Pull request by erwinmombay**

### CODE REVIEW

**Comment by erwinmombay** [comment]
@cramforce we ok with something like this?

screen shot 2016-04-22 at 6 16 51 pm

**Comment by erwinmombay** [comment]
might expand it to be a deeper graph for better error reporting and just add `find` helpers

File: build-system/tasks/jsify-css.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/**
```
Comment (cramforce) [inline build-system/tasks/jsify-css.js]: Wanna make this a separate PR? Or put it into the description.

File: build-system/tasks/dep-check.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          .map((x) => x.replace(`${root}/`, ''));
+      modules[x.name] = deps;
+    })
+    return modules;
+}
+
+/**
+ * @param {!Object<string, !Array>} modules
+ */
+function checkImplImport(modules) {
```
Comment (cramforce) [inline build-system/tasks/dep-check.js]: This PR does the heavy lifting and we can submit it like this, but what I'm hoping for is to have a declarative way to configure forbidden deps.

E.g.

```js
const rules = [
  {
    type: 'forbidden',
    filesMatching: 'extensions/**/*.js',
    mustNotDependOn: 'src/services/**/*.js',
    whitelist: [
      ['foo/bar.js', 'src/service/cid.js']
    ]
  }
]
```

Comment (erwinmombay) [inline build-system/tasks/dep-check.js]: yep. just havent added it. i have some code already i havent pushed yet that checks for any direct "-impl.js" imports. ill add that rule scheme in. shouldnt be hard

File: build-system/tasks/jsify-css.js (hunk 1 above)
Comment (erwinmombay) [inline build-system/tasks/jsify-css.js]: i can split it into a separate PR. thought i would need it but browserify didnt blow up for some reason w/o the CSS dependency being there. will investigate

File: build-system/tasks/dep-check.js (hunk 2 above)
Comment (erwinmombay) [inline build-system/tasks/dep-check.js]: @cramforce PTAL. added rule class and rule object config. intentionally red right now since amp-analytics.js directly import services files

**Comment by erwinmombay** [comment]
error on travis:

screen shot 2016-04-25 at 1 31 27 am

**Comment by erwinmombay** [comment]
added https://github.com/ampproject/amphtml/issues/2988 for improvement to consolidate the browserify run with build task.

File: build-system/dep-check-config.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * @typedef {{
+ *   type: (string|undefined),
+ *   filesMatching: (string|!Array<string>|undefined),
+ *   mustNotDependOn: (string|!Array<string>|undefined),
+ *   whitelist: (string|!Array<string>|undefined),
+ * }}
+ */
+var RuleConfigDef;
+
+exports.rules = [
```
Comment (cramforce) [inline build-system/dep-check-config.js]: <3

**Comment by cramforce** [comment]
Super awesome!

LGTM

File: build-system/dep-check-config.js (hunk 3 above)
Comment (erwinmombay) [inline build-system/dep-check-config.js]: @cramforce will need to whitelist analytics to get green build. is it intentional that we are importing the services directly there? doesnt seem right. https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/amp-analytics.js#L22

Comment (cramforce) [inline build-system/dep-check-config.js]: It is intentional! The analytics package provides that service.

Basically we want a forbidden deps test from everywhere else to that cid file
